selfdrive/updated.py

```diff
@@ -303,11 +305,36 @@ def attempt_update():
 
     finalize_from_ovfs_copy()
 
+    # If a NEOS update is required, download it in the background
+    with open("/VERSION", "r") as current_neos_file:
+      current_neos_version = current_neos_file.read().replace("\n", "")
+    required_neos_version = run(["bash", "-c", 
+               r"unset REQUIRED_NEOS_VERSION && source launch_env.sh && echo -n $REQUIRED_NEOS_VERSION"],
+               FINALIZED)
+    print(f"NEOS version update check: {current_neos_version} current, {required_neos_version} in update")
+    if current_neos_version != required_neos_version or DEBUG_FORCE_UPDATE:
+      print(f"Beginning background download for NEOS {required_neos_version}")
+      update_json = f'file:///{FINALIZED}/installer/updater/update.json'
+      while True:
+        Params().put("Offroad_NeosUpdate", "1")
+        try:
+          run(NICE_LOW_PRIORITY + ["installer/updater/updater", "bgcache", update_json], FINALIZED)
+          Params().put("Offroad_NeosUpdate", "0")
+          print("NEOS background download successful!")
+          break
+        except subprocess.CalledProcessError:
```

Comment:
What happens if the file is corrupted? Running the updater again is not going to fix it. The corrupted file will only get cleared on reboot.

Reply:
The NEOS updater will unlink() a file if it fails the signature check:

https://github.com/commaai/openpilot/blob/c43df81214d34d90b318defbd7f43e240b0c8f45/installer/updater/updater.cc#L332-L339

Historically the user gets to reboot and try again at that point. In the background cache mode, the NEOS updater still bails out, but the openpilot updater will call it again later and get a fresh try at it.

Reply:
Ah, missed the `unlink`